File: 3776-minimum-moves-to-balance-circular-array/3776-minimum-moves-to-balance-circular-array.py
class Solution:
   def minMoves(self, balance):
    n = len(balance)
    total = sum(balance)
    if total < 0:
        return -1

    # find the negative index
    neg = -1
    for i in range(n):
        if balance[i] < 0:
            neg = i
            break

    # already valid
    if neg == -1:
        return 0

    need = -balance[neg]
    suppliers = []

    for i in range(n):
        if i == neg or balance[i] == 0:
            continue
        dist = min((i - neg) % n, (neg - i) % n)
        suppliers.append((dist, balance[i]))

    suppliers.sort()

    moves = 0
    for dist, supply in suppliers:
        take = min(supply, need)
        moves += take * dist
        need -= take
        if need == 0:
            break

    return moves

File: 3776-minimum-moves-to-balance-circular-array/3776-minimum-moves-to-balance-circular-array.cpp
class Solution {
public:
    long long minMoves(vector<int>& balance) {
        int n=balance.size();
        long long totalMoves=0;
        long long sum=accumulate(balance.begin(),balance.end(),0ll);
        if(sum<0) return -1;
        int negativeIndex=-1;
        for(int i=0;i<n;i++){
            if(balance[i]<0){
                negativeIndex=i;
            }
        }
        if(negativeIndex==-1) return 0;
        int distance=1;
        while(balance[negativeIndex]<0){
            int leftIndex=(negativeIndex-distance+n)%n;
            int rightIndex=(negativeIndex+distance)%n;
            long long leftRightSum=balance[leftIndex]+balance[rightIndex];
            long long need = -1*balance[negativeIndex];
            balance[negativeIndex]+=min(need,leftRightSum);
            totalMoves+=distance*min(need,leftRightSum);
            distance++;
        }
        return totalMoves;
    }
};